File: writer.py
"""Output file writer and ffmpeg audio fixer."""

import os
import re
import struct
import shutil
import subprocess


def write_output(corrupted_path, output_path, scan_info, moov):
    """Write recovered MP4: copy pre-mdat + mdat data + append moov."""
    mdat_offset = scan_info['mdat_offset']
    mdat_size = scan_info['mdat_size']

    with open(corrupted_path, 'rb') as src, open(output_path, 'wb') as dst:
        # Copy pre-mdat boxes (ftyp, uuid, etc.)
        src.seek(0)
        dst.write(src.read(mdat_offset))

        # Detect original mdat header size (8 standard, 16 extended)
        src.seek(mdat_offset)
        orig_size_field = struct.unpack('>I', src.read(4))[0]
        orig_hdr_size = 16 if orig_size_field == 1 else 8

        # Write mdat with correct size header
        if mdat_size > 0xFFFFFFFF:
            dst.write(struct.pack('>I4sQ', 1, b'mdat', mdat_size))
            src.seek(mdat_offset + orig_hdr_size)
            remaining = mdat_size - orig_hdr_size
        else:
            dst.write(struct.pack('>I4s', mdat_size, b'mdat'))
            src.seek(mdat_offset + orig_hdr_size)
            remaining = mdat_size - orig_hdr_size

        # Stream mdat content
        BUF = 4 * 1024 * 1024
        while remaining > 0:
            chunk = src.read(min(BUF, remaining))
            if not chunk:
                break
            dst.write(chunk)
            remaining -= len(chunk)

        # Append moov
        dst.write(moov)


def fix_audio(output_path):
    """Fix audio: validate AAC stream, re-encode only if needed."""
    ffmpeg = shutil.which('ffmpeg')
    if not ffmpeg:
        print("  WARNING: ffmpeg not found - skipping audio fix.")
        print("  Run manually: ffmpeg -i OUTPUT -c:v copy -c:a aac -b:a 192k FIXED.mp4")
        return False

    # Check if audio decodes cleanly (lossless path)
    import time as _time
    print("  Validating audio stream...", end='', flush=True)
    t_chk = _time.perf_counter()
    r = subprocess.run(
        [ffmpeg, '-v', 'error', '-i', output_path, '-vn', '-f', 'null', '-'],
        capture_output=True, text=True, timeout=600)
    aac_errors = [l for l in (r.stderr or '').split('\n')
                  if l.strip() and 'aac' in l.lower()]
    chk_elapsed = _time.perf_counter() - t_chk
    if not aac_errors:
        print(f" clean. ({chk_elapsed:.1f}s)")
        return True

    print(f" {len(aac_errors)} AAC errors. ({chk_elapsed:.1f}s)")
    print(f"  Re-encoding audio...")
    return _fix_audio_reencode(output_path, ffmpeg)


def _print_seg_progress(done, total, lossless, reencoded, silent, t0, action):
    """Print a single-line progress bar for segment processing."""
    import time as _time
    pct = 100 * done / total if total else 0
    filled = int(30 * done / total) if total else 0
    bar = '#' * filled + '-' * (30 - filled)
    elapsed = _time.perf_counter() - t0
    eta = (elapsed / done * (total - done)) if done > 0 else 0
    status = f"L:{lossless} R:{reencoded} S:{silent}"
    print(f"\r  [{bar}] {done}/{total} ({pct:.0f}%) "
          f"{status} [{action}] "
          f"elapsed:{elapsed:.0f}s eta:{eta:.0f}s   ",
          end='', flush=True)


def _fix_audio_reencode(output_path, ffmpeg):
    """Hybrid audio fix: lossless copy for clean segments, re-encode corrupt ones.

    Optimised pipeline:
    1. Extract full audio track once (copy) — avoids repeated seeks into large file
    2. Extract + test segments in parallel from the small audio file
    3. Re-encode only bad segments in parallel
    4. Concatenate all segments and mux back with original video
    """
    import tempfile
    import time as _time
    from concurrent.futures import ThreadPoolExecutor, as_completed

    seg_dur = 2  # seconds per segment (small to maximize lossless)

    # Get total duration
    r = subprocess.run(
        [ffmpeg, '-i', output_path],
        capture_output=True, text=True, timeout=30)
    duration = 0.0
    for line in (r.stderr or '').split('\n'):
        m = re.search(r'Duration:\s*(\d+):(\d+):(\d+)\.(\d+)', line)
        if m:
            duration = (int(m.group(1)) * 3600 + int(m.group(2)) * 60 +
                        int(m.group(3)) + int(m.group(4)) / 100)
            break
    if duration <= 0:
        print("  WARNING: Could not determine duration.")
        return False

    total_segs = int(duration / seg_dur) + (1 if duration % seg_dur else 0)
    workers = min(8, os.cpu_count() or 4)
    tmpdir = tempfile.mkdtemp(prefix='recover_mp4_')
    t0 = _time.perf_counter()

    try:
        # Phase 1: Extract full audio track (fast stream-copy from source)
        print("  Extracting audio track...", end='', flush=True)
        full_audio = os.path.join(tmpdir, 'full_audio.m4a')
        r = subprocess.run(
            [ffmpeg, '-y', '-v', 'error', '-i', output_path,
             '-vn', '-c:a', 'copy', full_audio],
            capture_output=True, text=True, timeout=120)
        if r.returncode != 0:
            print(f" failed.")
            return False
        print(" done.")

        # Phase 2: Extract + test segments in parallel (from small audio file)
        seg_files = [os.path.join(tmpdir, f'seg_{i:05d}.m4a')
                     for i in range(total_segs)]

        def extract_and_test(i):
            """Extract segment from full_audio and test for AAC errors."""
            ss = i * seg_dur
            t = min(seg_dur, duration - ss)
            seg_path = seg_files[i]
            r = subprocess.run(
                [ffmpeg, '-y', '-v', 'error',
                 '-ss', f'{ss:.3f}', '-t', f'{t:.3f}',
                 '-i', full_audio, '-vn', '-c:a', 'copy', seg_path],
                capture_output=True, text=True, timeout=60)
            if r.returncode != 0 or not os.path.exists(seg_path):
                return 'bad'
            r2 = subprocess.run(
                [ffmpeg, '-v', 'error', '-i', seg_path, '-f', 'null', '-'],
                capture_output=True, text=True, timeout=60)
            errs = [l for l in (r2.stderr or '').split('\n')
                    if l.strip() and 'aac' in l.lower()]
            return 'bad' if errs else 'good'

        bad_indices = set()
        done = 0
        with ThreadPoolExecutor(max_workers=workers) as pool:
            futures = {pool.submit(extract_and_test, i): i
                       for i in range(total_segs)}
            for future in as_completed(futures):
                if future.result() == 'bad':
                    bad_indices.add(futures[future])
                done += 1
                if done % 20 == 0 or done == total_segs:
                    _print_seg_progress(done, total_segs,
                                        done - len(bad_indices), 0, 0,
                                        t0, 'testing')

        # Phase 3: Re-encode bad segments in parallel
        reencoded = 0
        silent = 0
        if bad_indices:
            print(f"\n  Re-encoding {len(bad_indices)} bad segments...",
                  end='', flush=True)

        def fix_seg(i):
            """Re-encode a bad segment; fall back to silence on failure."""
            ss = i * seg_dur
            t = min(seg_dur, duration - ss)
            seg_path = seg_files[i]
            tmp_path = seg_path + '.tmp.m4a'
            r = subprocess.run(
                [ffmpeg, '-y', '-v', 'error',
                 '-err_detect', 'ignore_err',
                 '-fflags', '+genpts+discardcorrupt',
                 '-ss', f'{ss:.3f}', '-t', f'{t:.3f}',
                 '-i', full_audio,
                 '-c:a', 'aac', '-ac', '2', '-b:a', '192k', tmp_path],
                capture_output=True, text=True, timeout=120)
            if r.returncode == 0 and os.path.exists(tmp_path):
                os.replace(tmp_path, seg_path)
                return 'reencoded'
            if os.path.exists(tmp_path):
                try:
                    os.remove(tmp_path)
                except OSError:
                    pass
            subprocess.run(
                [ffmpeg, '-y', '-v', 'error',
                 '-f', 'lavfi', '-i', 'anullsrc=r=48000:cl=stereo',
                 '-t', f'{t:.3f}',
                 '-c:a', 'aac', '-b:a', '192k', seg_path],
                capture_output=True, text=True, timeout=30)
            return 'silent'

        if bad_indices:
            t_fix = _time.perf_counter()
            with ThreadPoolExecutor(max_workers=workers) as pool:
                futures = {pool.submit(fix_seg, i): i for i in bad_indices}
                for future in as_completed(futures):
                    result = future.result()
                    if result == 'reencoded':
                        reencoded += 1
                    else:
                        silent += 1
            fix_elapsed = _time.perf_counter() - t_fix
            print(f" done. ({fix_elapsed:.1f}s)")

        lossless = total_segs - len(bad_indices)
        print(f"\r  Segments: {lossless}/{total_segs} lossless, "
              f"{reencoded}/{total_segs} re-encoded, "
              f"{silent}/{total_segs} silence.{' ' * 20}")

        # Phase 4: Concatenate all segments
        print("  Concatenating segments...", end='', flush=True)
        concat_list = os.path.join(tmpdir, 'concat.txt')
        with open(concat_list, 'w') as f:
            for seg in seg_files:
                f.write(f"file '{seg}'\n")

        concat_audio = os.path.join(tmpdir, 'concat_audio.m4a')
        r = subprocess.run(
            [ffmpeg, '-y', '-v', 'error',
             '-f', 'concat', '-safe', '0', '-i', concat_list,
             '-c:a', 'copy', concat_audio],
            capture_output=True, text=True, timeout=300)
        if r.returncode != 0:
            print(f"\n  WARNING: Concat failed: {(r.stderr or '')[-200:]}")
            return False
        print(" done.")

        # Phase 5: Mux original video + fixed audio
        print("  Muxing video + fixed audio...", end='', flush=True)
        tmp = output_path + '.tmp.mp4'
        r = subprocess.run(
            [ffmpeg, '-y', '-v', 'error',
             '-i', output_path, '-i', concat_audio,
             '-map', '0:v', '-map', '1:a',
             '-c:v', 'copy', '-c:a', 'copy',
             '-movflags', '+faststart', tmp],
            capture_output=True, text=True, timeout=3600)
        if r.returncode == 0 and os.path.exists(tmp):
            os.replace(tmp, output_path)
            elapsed = _time.perf_counter() - t0
            print(f" done. ({elapsed:.1f}s)")
            print("  Audio fixed.")
            return True
        print(f"\n  WARNING: Final mux failed: {(r.stderr or '')[-200:]}")
        if os.path.exists(tmp):
            try:
                os.remove(tmp)
            except OSError:
                pass
        return False
    finally:
        import shutil as _shutil
        _shutil.rmtree(tmpdir, ignore_errors=True)

File: cpp/src/writer.cpp
/// Output file writer and ffmpeg subprocess audio fixer.
///
/// write_output: streams mdat from mmap'd source to output file, appends moov.
/// fix_audio: hybrid parallel audio fix — extract once, test/fix segments in parallel.

#include "writer.hpp"
#include "constants.hpp"
#include "mmap_file.hpp"

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <format>
#include <fstream>
#include <future>
#include <print>
#include <regex>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

namespace recover {

void write_output(const std::string& corrupted_path, const std::string& output_path,
                  const ScanResult& scan, const std::vector<uint8_t>& moov) {

    MappedFile src(corrupted_path);
    const uint8_t* data = src.data();

    std::ofstream dst(output_path, std::ios::binary);
    if (!dst)
        throw std::runtime_error("Cannot open output file: " + output_path);

    // Copy pre-mdat boxes (ftyp, uuid, etc.)
    dst.write(reinterpret_cast<const char*>(data), scan.mdat_offset);

    // Detect original mdat header size
    uint32_t orig_size_field = read_be32(data + scan.mdat_offset);
    size_t orig_hdr_size = (orig_size_field == 1) ? 16 : 8;

    // Write mdat header.
    // data_bytes = actual payload bytes (mdat_size includes the original header).
    uint64_t data_bytes = scan.mdat_size - orig_hdr_size;
    if (scan.mdat_size > 0xFFFFFFFF) {
        // Extended 64-bit size: total box = 16-byte header + data_bytes.
        uint8_t hdr[16];
        write_be32(hdr, 1);
        std::memcpy(hdr + 4, "mdat", 4);
        write_be64(hdr + 8, 16 + data_bytes);
        dst.write(reinterpret_cast<const char*>(hdr), 16);
    } else {
        uint8_t hdr[8];
        write_be32(hdr, static_cast<uint32_t>(orig_hdr_size + data_bytes));
        std::memcpy(hdr + 4, "mdat", 4);
        dst.write(reinterpret_cast<const char*>(hdr), 8);
    }

    // Copy mdat content (skip original header)
    const uint8_t* mdat_content = data + scan.mdat_offset + orig_hdr_size;
    dst.write(reinterpret_cast<const char*>(mdat_content), static_cast<std::streamsize>(data_bytes));

    // Append moov
    dst.write(reinterpret_cast<const char*>(moov.data()), moov.size());
}


// ─── Cross-platform subprocess helpers ──────────────────────────────────────

namespace {

/// Run ffmpeg command and capture stderr. Returns exit code.
int run_ffmpeg(const std::vector<std::string>& args, std::string* captured_stderr = nullptr) {
    std::string cmd;
    for (const auto& a : args) {
        if (!cmd.empty()) cmd += ' ';
        // Quote arguments containing spaces or special chars
        if (a.find(' ') != std::string::npos || a.find('\'') != std::string::npos) {
            cmd += '"';
            cmd += a;
            cmd += '"';
        } else {
            cmd += a;
        }
    }

    // Redirect stderr to a temp file if we need to capture it
    std::string stderr_file;
    if (captured_stderr) {
        static std::atomic<uint64_t> stderr_counter{0};
        auto tmp = std::filesystem::temp_directory_path() /
                   std::format("ffmpeg_err_{}_{}.txt",
                               std::this_thread::get_id(),
                               stderr_counter.fetch_add(1, std::memory_order_relaxed));
        stderr_file = tmp.string();
        cmd += " 2>\"" + stderr_file + "\"";
    } else {
#ifdef _WIN32
        cmd += " 2>NUL";
#else
        cmd += " 2>/dev/null";
#endif
    }
#ifdef _WIN32
    cmd += " >NUL";
#else
    cmd += " >/dev/null";
#endif

    int ret = std::system(cmd.c_str());

    if (captured_stderr && !stderr_file.empty()) {
        std::ifstream ifs(stderr_file);
        if (ifs) {
            *captured_stderr = std::string(
                std::istreambuf_iterator<char>(ifs),
                std::istreambuf_iterator<char>());
        }
        std::error_code ec;
        std::filesystem::remove(stderr_file, ec);
    }

    return ret;
}

/// Probe file duration in seconds using ffmpeg.
double probe_duration(const std::string& path) {
    std::string output;
    run_ffmpeg({"ffmpeg", "-i", path}, &output);

    std::regex dur_re(R"(Duration:\s*(\d+):(\d+):(\d+)\.(\d+))");
    std::smatch m;
    if (std::regex_search(output, m, dur_re)) {
        return std::stoi(m[1]) * 3600.0 + std::stoi(m[2]) * 60.0 +
               std::stoi(m[3]) + std::stoi(m[4]) / 100.0;
    }
    return 0.0;
}

/// Count AAC errors in ffmpeg stderr output.
int count_aac_errors(const std::string& stderr_output) {
    int count = 0;
    std::istringstream iss(stderr_output);
    std::string line;
    while (std::getline(iss, line)) {
        if (line.empty()) continue;
        std::string lower = line;
        std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
        if (lower.find("aac") != std::string::npos) {
            ++count;
        }
    }
    return count;
}

/// Remove file if it exists (no-throw).
void try_remove(const std::string& path) {
    std::error_code ec;
    std::filesystem::remove(path, ec);
}

/// Print progress bar for segment processing.
void print_progress(int done, int total, int lossless, int reencoded, int silent,
                    double elapsed, const std::string& action) {
    double pct = (total > 0) ? 100.0 * done / total : 0;
    int filled = (total > 0) ? 30 * done / total : 0;
    double eta = (done > 0) ? elapsed / done * (total - done) : 0;
    std::string bar(filled, '#');
    bar.append(30 - filled, '-');
    std::print("\r  [{}] {}/{} ({:.0f}%) L:{} R:{} S:{} [{}] "
               "elapsed:{:.0f}s eta:{:.0f}s   ",
               bar, done, total, pct, lossless, reencoded, silent,
               action, elapsed, eta);
    std::fflush(stdout);
}

} // anonymous namespace


bool fix_audio(const std::string& output_path) {
    // Check if ffmpeg is available
#ifdef _WIN32
    if (std::system("ffmpeg -version >NUL 2>NUL") != 0) {
#else
    if (std::system("ffmpeg -version >/dev/null 2>/dev/null") != 0) {
#endif
        std::println("  WARNING: ffmpeg not found — skipping audio fix.");
        std::println("  Run manually: ffmpeg -i OUTPUT -c:v copy -c:a aac -b:a 192k FIXED.mp4");
        return false;
    }

    // Check if audio decodes cleanly
    std::print("  Validating audio stream...");
    std::fflush(stdout);
    auto t_chk = std::chrono::steady_clock::now();
    std::string stderr_out;
    run_ffmpeg({"ffmpeg", "-v", "error", "-i", output_path, "-vn", "-f", "null", "-"},
               &stderr_out);
    int aac_errors = count_aac_errors(stderr_out);
    double chk_elapsed = std::chrono::duration<double>(
        std::chrono::steady_clock::now() - t_chk).count();
    if (aac_errors == 0) {
        std::println(" clean. ({:.1f}s)", chk_elapsed);
        return true;
    }
    std::println(" {} AAC errors. ({:.1f}s)", aac_errors, chk_elapsed);
    std::println("  Re-encoding audio...");

    // ── Hybrid parallel audio fix ──
    auto t0 = std::chrono::steady_clock::now();

    double duration = probe_duration(output_path);
    if (duration <= 0) {
        std::println("  WARNING: Could not determine duration.");
        return false;
    }

    constexpr double seg_dur = 2.0;
    int total_segs = static_cast<int>(duration / seg_dur) +
                     ((std::fmod(duration, seg_dur) > 0) ? 1 : 0);
    int workers = std::min(8, static_cast<int>(std::thread::hardware_concurrency()));
    if (workers < 1) workers = 4;

    auto tmpdir = std::filesystem::temp_directory_path() / "recover_mp4_cpp";
    std::error_code ec;
    std::filesystem::create_directories(tmpdir, ec);

    auto cleanup = [&]() {
        std::error_code ec2;
        std::filesystem::remove_all(tmpdir, ec2);
    };

    // Phase 1: Extract full audio track
    std::print("  Extracting audio track...");
    std::fflush(stdout);
    std::string full_audio = (tmpdir / "full_audio.m4a").string();
    int ret = run_ffmpeg({"ffmpeg", "-y", "-v", "error", "-i", output_path,
                          "-vn", "-c:a", "copy", full_audio});
    if (ret != 0) {
        std::println(" failed.");
        cleanup();
        return false;
    }
    std::println(" done.");

    // Build segment file paths
    std::vector<std::string> seg_files(total_segs);
    for (int i = 0; i < total_segs; ++i) {
        seg_files[i] = (tmpdir / std::format("seg_{:05d}.m4a", i)).string();
    }

    // Phase 2: Extract + test segments in parallel
    auto extract_and_test = [&](int i) -> bool {
        double ss = i * seg_dur;
        double t = std::min(seg_dur, duration - ss);
        const auto& seg_path = seg_files[i];

        int r = run_ffmpeg({"ffmpeg", "-y", "-v", "error",
                            "-ss", std::format("{:.3f}", ss),
                            "-t", std::format("{:.3f}", t),
                            "-i", full_audio, "-vn", "-c:a", "copy", seg_path});
        if (r != 0 || !std::filesystem::exists(seg_path)) return false;

        std::string err;
        run_ffmpeg({"ffmpeg", "-v", "error", "-i", seg_path, "-f", "null", "-"}, &err);
        return count_aac_errors(err) == 0;
    };

    std::vector<bool> seg_good(total_segs, false);
    {
        std::vector<std::future<bool>> futures;
        int done = 0;
        for (int start = 0; start < total_segs; start += workers) {
            int batch_end = std::min(start + workers, total_segs);
            futures.clear();
            for (int i = start; i < batch_end; ++i) {
                futures.push_back(std::async(std::launch::async, extract_and_test, i));
            }
            for (int i = 0; i < static_cast<int>(futures.size()); ++i) {
                seg_good[start + i] = futures[i].get();
                ++done;
                if (done % 20 == 0 || done == total_segs) {
                    auto now = std::chrono::steady_clock::now();
                    double elapsed = std::chrono::duration<double>(now - t0).count();
                    int good_count = 0;
                    for (int j = 0; j < done; ++j) if (seg_good[j]) ++good_count;
                    print_progress(done, total_segs, good_count, 0, 0, elapsed, "testing");
                }
            }
        }
    }

    // Phase 3: Re-encode bad segments in parallel
    std::vector<int> bad_indices;
    for (int i = 0; i < total_segs; ++i) {
        if (!seg_good[i]) bad_indices.push_back(i);
    }

    std::atomic<int> reencoded{0}, silent_count{0};

    auto fix_seg = [&](int i) {
        double ss = i * seg_dur;
        double t = std::min(seg_dur, duration - ss);
        const auto& seg_path = seg_files[i];
        std::string tmp_path = seg_path + ".tmp.m4a";

        int r = run_ffmpeg({"ffmpeg", "-y", "-v", "error",
                            "-err_detect", "ignore_err",
                            "-fflags", "+genpts+discardcorrupt",
                            "-ss", std::format("{:.3f}", ss),
                            "-t", std::format("{:.3f}", t),
                            "-i", full_audio,
                            "-c:a", "aac", "-ac", "2", "-b:a", "192k", tmp_path});
        if (r == 0 && std::filesystem::exists(tmp_path)) {
            std::filesystem::rename(tmp_path, seg_path);
            ++reencoded;
            return;
        }
        try_remove(tmp_path);
        // Silence fallback
        run_ffmpeg({"ffmpeg", "-y", "-v", "error",
                    "-f", "lavfi", "-i", "anullsrc=r=48000:cl=stereo",
                    "-t", std::format("{:.3f}", t),
                    "-c:a", "aac", "-b:a", "192k", seg_path});
        ++silent_count;
    };

    if (!bad_indices.empty()) {
        std::print("\n  Re-encoding {} bad segments...", bad_indices.size());
        std::fflush(stdout);
        auto t_fix = std::chrono::steady_clock::now();
        std::vector<std::future<void>> futures;
        for (int start = 0; start < static_cast<int>(bad_indices.size()); start += workers) {
            int batch_end = std::min(start + workers, static_cast<int>(bad_indices.size()));
            futures.clear();
            for (int i = start; i < batch_end; ++i) {
                futures.push_back(std::async(std::launch::async, fix_seg, bad_indices[i]));
            }
            for (auto& f : futures) f.get();
        }
        double fix_elapsed = std::chrono::duration<double>(
            std::chrono::steady_clock::now() - t_fix).count();
        std::println(" done. ({:.1f}s)", fix_elapsed);
    }

    int lossless = total_segs - static_cast<int>(bad_indices.size());
    std::println("\r  Segments: {}/{} lossless, {}/{} re-encoded, "
                 "{}/{} silence.{:>20}",
                 lossless, total_segs, reencoded.load(), total_segs,
                 silent_count.load(), total_segs, "");

    // Phase 4: Concatenate all segments
    std::print("  Concatenating segments...");
    std::fflush(stdout);
    std::string concat_list = (tmpdir / "concat.txt").string();
    {
        std::ofstream cl(concat_list);
        for (const auto& seg : seg_files) {
            cl << "file '" << seg << "'\n";
        }
    }
    std::string concat_audio = (tmpdir / "concat_audio.m4a").string();
    ret = run_ffmpeg({"ffmpeg", "-y", "-v", "error",
                      "-f", "concat", "-safe", "0", "-i", concat_list,
                      "-c:a", "copy", concat_audio});
    if (ret != 0) {
        std::println(" failed.");
        cleanup();
        return false;
    }
    std::println(" done.");

    // Phase 5: Mux original video + fixed audio
    std::print("  Muxing video + fixed audio...");
    std::fflush(stdout);
    std::string tmp = output_path + ".tmp.mp4";
    ret = run_ffmpeg({"ffmpeg", "-y", "-v", "error",
                      "-i", output_path, "-i", concat_audio,
                      "-map", "0:v", "-map", "1:a",
                      "-c:v", "copy", "-c:a", "copy",
                      "-movflags", "+faststart", tmp});
    if (ret == 0 && std::filesystem::exists(tmp)) {
        std::filesystem::rename(tmp, output_path);
        auto now = std::chrono::steady_clock::now();
        double elapsed = std::chrono::duration<double>(now - t0).count();
        std::println(" done. ({:.1f}s)", elapsed);
        std::println("  Audio fixed.");
        cleanup();
        return true;
    }

    std::println(" failed.");
    try_remove(tmp);
    cleanup();
    return false;
}

} // namespace recover
